File: app/src/main/java/com/jinhanexample/imageSearchApp/di/AppModule.java
package com.jinhanexample.imageSearchApp.di;

import dagger.Module;
import dagger.hilt.InstallIn;
import dagger.hilt.android.components.ApplicationComponent;

//ApplicationComponent는 Application 전체의 생명주기를 lifetime으로 갖습니다.
// Application이 생성되는(onCreate) 시점에 함께 생성되고, Application이 파괴되는(onDestroy) 시점에 함께 파괴됩니다.
//Hilt는 표준적으로 제공하는 component 들이 이미 존재하기 때문에
// @InstallIn 어노테이션을 사용하여 표준 component에 module들을 install 할 수 있습니다.
// Hilt에서 제공하는 기본적인 규칙은 모든 module에 @InstallIn 어노테이션을 사용하여 어떤 component에 install 할지 반드시 정해주어야 합니다
@Module
@InstallIn(ApplicationComponent.class)
public class AppModule {
}

File: app/src/main/java/com/jinhanexample/mvvmTodo/di/AppModule.kt
package com.jinhanexample.mvvmTodo.di

import android.app.Application
import androidx.room.Room
import com.jinhanexample.mvvmTodo.data.TaskDatabase
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.components.ApplicationComponent
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import javax.inject.Qualifier
import javax.inject.Singleton

//Module은 객체(object)에 붙인다.
@Module
@InstallIn(ApplicationComponent::class)
object AppModule {

    //provides는 메소드에 붙인다.
    //database를 생성하는 메소드
    //파라미터로 받은 application에 TaskDatabase를 생성한다.
    @Provides
    @Singleton
    fun provideDatabase(
        app : Application,
        callback : TaskDatabase.Callback
    ) =
        Room.databaseBuilder(app, TaskDatabase::class.java, "task_database")
            .fallbackToDestructiveMigration()
            .addCallback(callback)
            .build()

    @Provides
    fun provideTaskDao(db : TaskDatabase) = db.taskDao()

    @ApplicationScope
    @Provides
    @Singleton
    fun provideApplicationScope() = CoroutineScope(SupervisorJob())
}

//Annotation 만드는 방법
@Retention(AnnotationRetention.RUNTIME)
@Qualifier
annotation class ApplicationScope